snakemake/wrapper.py

```diff
@@ -1,18 +1,204 @@
 __author__ = "Johannes Köster"
-__copyright__ = "Copyright 2016-2019, Johannes Köster"
+__copyright__ = "Copyright 2016-2020, Johannes Köster"
 __email__ = "koester@jimmy.harvard.edu"
 __license__ = "MIT"
 
 
 import os
 import posixpath
+import re
+import hashlib
+import requests
+import yaml
 
 from urllib.error import URLError
 from urllib.request import urlopen
 
+from snakemake.deployment.docker import DockerContainer
 from snakemake.exceptions import WorkflowError
 from snakemake.script import script
 
+# Derive custom container prefix from the environment, default to quay.io
+CONTAINER_PREFIX = os.environ.get(
+    "SNAKEMAKE_WRAPPER_CONTAINER_PREFIX", "quay.io/snakemake-wrappers"
+)
+WRAPPER_REPOSITORY = os.environ.get(
+    "SNAKEMAKE_WRAPPER_REPOSITORY",
+    "https://raw.githubusercontent.com/snakemake/snakemake-wrappers/",
+)
+
+
+class WrapperSpec:
```

Comment:
Here is the WrapperSpec! I had trouble writing this because I don't fully understand all the options that might be given for a path for a snakemake wrapper (e.g., git+file? file:?). I think the logic here can be hugely simplified to not have so many supporting functions - but I don't understand it enough to do it.